mava/wrappers/gym.py

```diff
@@ -0,0 +1,312 @@
+    ) -> Tuple[np.ndarray, Dict]:
+
+        # Reset the env
+        agents_view, info = self._env.reset(seed, options)
+
+        # Create the metrics dict
+        metrics = {
+            "episode_return": self.running_count_episode_return,
+            "episode_length": self.running_count_episode_length,
+            "is_terminal_step": True,
+        }
+
+        # Reset the metrics
+        self.running_count_episode_return = 0.0
+        self.running_count_episode_length = 0
+
+        if "won_episode" in info:
+            metrics["won_episode"] = info["won_episode"]
+
+        info["metrics"] = metrics
+
+        return agents_view, info
+
+    def step(self, actions: NDArray) -> Tuple:
+
+        # Step the env
+        agents_view, reward, terminated, truncated, info = self._env.step(actions)
+
+        self.running_count_episode_return += float(np.mean(reward))
+        self.running_count_episode_length += 1
+
+        metrics = {
+            "episode_return": self.running_count_episode_return,
+            "episode_length": self.running_count_episode_length,
+            "is_terminal_step": False,
+        }
+        if "won_episode" in info:
+            metrics["won_episode"] = info["won_episode"]
+
+        info["metrics"] = metrics
+
+        return agents_view, reward, terminated, truncated, info
+
+
+class GymAgentIDWrapper(gym.Wrapper):
+    """Add one hot agent IDs to observation."""
+
+    def __init__(self, env: gym.Env):
+        super().__init__(env)
+
+        self.agent_ids = np.eye(self.env.num_agents)
+        observation_space = self.env.observation_space[0]
+        _obs_low, _obs_high, _obs_dtype, _obs_shape = (
+            observation_space.low[0],
+            observation_space.high[0],
+            observation_space.dtype,
+            observation_space.shape,
+        )
+        _new_obs_shape = (_obs_shape[0] + self.env.num_agents,)
+        _observation_boxs = [
+            spaces.Box(low=_obs_low, high=_obs_high, shape=_new_obs_shape, dtype=_obs_dtype)
+        ] * self.env.num_agents
+        self.observation_space = spaces.Tuple(_observation_boxs)
+
+    def reset(
+        self, seed: Optional[int] = None, options: Optional[dict] = None
+    ) -> Tuple[np.ndarray, Dict]:
+        """Reset the environment."""
+        obs, info = self.env.reset(seed, options)
+        obs = np.concatenate([self.agent_ids, obs], axis=1)
+        return obs, info
+
+    def step(self, action: list) -> Tuple[np.ndarray, float, bool, bool, Dict]:
+        """Step the environment."""
+        obs, reward, terminated, truncated, info = self.env.step(action)
+        obs = np.concatenate([self.agent_ids, obs], axis=1)
+        return obs, reward, terminated, truncated, info
+
+
+# Copied form https://github.com/openai/gym/blob/master/gym/vector/async_vector_env.py
+# Modified to work with multiple agents
+def _multiagent_worker_shared_memory(  # noqa: CCR001
```

Comment:
may I ask why we need this, does not async automatically handling memory and autoresetting?

Reply:
I really don't like that we need to maintain this. Does pufferlib's wrapper not solve the multi agent/single agent issue?

Reply:
Yes, but i needed to change this line https://github.com/instadeepai/Mava/blob/133a25060151ccd99b0f0fe1a73af48310dbbbff/mava/wrappers/gym.py#L271
since originally it was:
```python3
if terminated or truncated:
```
leading to the environments resetting at each step when the terminated / truncated flags are returned as arrays.